### PR TITLE
Add opt-in memoization of component layout

### PR DESCRIPTION
**Pull request by darknoon**
This is the second half of the earlier memoization PR. 

Depends on https://github.com/facebook/componentkit/pull/269

### CODE REVIEW

**Comment by darknoon** [comment]
In actual use, I noticed that this oscillates between cached and uncached, since I don't store into the next cache when I return a cached component. Duh!

File: ComponentKit/Core/CKComponentMemoizer.h [hunk 1]
```diff
@@ -78,6 +89,9 @@ struct CKComponentMemoizer {
    */
   id nextMemoizerState();
 
+  // Copies the layout into the layout cache
+  static void rememberLayout(CKComponentLayout l);
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.h]: Delete? I think this shouldn't be here.

File: ComponentKit/Core/CKComponentMemoizer.h [hunk 2]
```diff
@@ -126,3 +140,8 @@ CKMemoizationKey CKMakeTupleMemoizationKey(Types... args) {
   return CKMemoizationKey{};
 };
 
+
+// Internal, for CKComponent
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.h]: /** Internal, for CKComponent */

File: ComponentKit/Core/CKComponentSubclass.h [hunk 3]
```diff
@@ -77,6 +77,11 @@ extern CGSize const kCKComponentParentSizeUndefined;
                       relativeToParentSize:(CGSize)parentSize;
 
 /**
+ TODO: document
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentSubclass.h]: Finish this up?

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 4]
```diff
@@ -29,6 +29,38 @@ size_t operator ()(CKMemoizationKey a) const {
...
+struct CKLayoutMemoizationKey {
+  CKComponent *component;
+  CKSizeRange thatFits;
+  CGSize parentSize;
+
+  struct Hash {
+    size_t operator ()(CKLayoutMemoizationKey a) const {
+      NSUInteger subhashes[] = {
+        a.component.hash,
+        std::hash<CGFloat>()(a.thatFits.min.width),
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: should probably just add a hash for CKComponentSize

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 5]
```diff
@@ -29,6 +29,38 @@ size_t operator ()(CKMemoizationKey a) const {
...
+      };
+      return CKIntegerArrayHash(subhashes, CK_ARRAY_COUNT(subhashes));
+    };
+  };
+
+  struct Equals {
+    bool operator ()(CKLayoutMemoizationKey a, CKLayoutMemoizationKey b) const {
+      return a.component == b.component
+      && a.thatFits == b.thatFits
+      && a.parentSize.width == b.parentSize.width
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: CGSizeEqualToSize?

**Comment by adamjernst** [comment]
Mostly good! Just a few things. Docs for shouldMemoize is probably most important.

File: ComponentKit/Core/CKComponentMemoizer.mm (hunk 4 above)
Comment (darknoon) [inline ComponentKit/Core/CKComponentMemoizer.mm]: That would be great once https://github.com/facebook/componentkit/pull/285 lands

File: ComponentKit/Core/CKComponentMemoizer.h (hunk 1 above)
Comment (darknoon) [inline ComponentKit/Core/CKComponentMemoizer.h]: Yeah, it was a legacy thing

**Pull request merged**